src/szk.rs

```diff
@@ -0,0 +1,444 @@
+    fn derive_helper_proofs_share(&self, proofs_share_seed: &Seed<SEED_SIZE>) -> Vec<T::Field> {
+        Prng::from_seed_stream(P::seed_stream(
+            proofs_share_seed,
+            &self.adjust_domsep_tag(DST_PROOF_SHARE),
+            &[0],
+        ))
+        .take(self.typ.proof_len())
+        .collect()
+    }
+
+    fn derive_query_rands(&self, verify_key: &[u8; SEED_SIZE], nonce: &[u8; 16]) -> Vec<T::Field> {
+        let mut xof = P::init(verify_key, &self.adjust_domsep_tag(DST_QUERY_RANDOMNESS));
+        xof.update(nonce);
+        xof.into_seed_stream()
+            .into_field_vec(self.typ.query_rand_len())
+    }
+
+    fn prove(
+        &self,
+        leader_input_share: &[T::Field],
+        helper_input_share: &[T::Field],
+        encoded_measurement: &[T::Field],
+        rand_seeds: [Seed<SEED_SIZE>; 2],
+        leader_seed_opt: Option<Seed<SEED_SIZE>>,
+        nonce: &[u8; 16],
+    ) -> Result<[SzkProofShare<SEED_SIZE, T::Field>; 2], SzkError> {
+        let [prove_rand_seed, helper_seed] = rand_seeds;
+        let (blind_and_joint_rand_part_for_leader, helper_joint_rand_part, joint_rand) =
+            if self.typ.joint_rand_len() > 0 {
+                let leader_seed = leader_seed_opt.clone().unwrap();
+                let leader_joint_rand_part =
+                    self.derive_joint_rand_part(&leader_seed, leader_input_share, nonce)?;
+                let helper_joint_rand_part =
+                    self.derive_joint_rand_part(&helper_seed, helper_input_share, nonce)?;
+                let (_jrs, jr) =
+                    self.derive_joint_rands(&leader_joint_rand_part, &helper_joint_rand_part);
+                (
+                    Some((leader_seed, helper_joint_rand_part)),
+                    Some(leader_joint_rand_part),
+                    jr,
+                )
+            } else {
+                (None, None, Vec::<T::Field>::new())
+            };
+
+        let prove_rand = self.derive_prove_rands(&prove_rand_seed);
+        let mut leader_proofs_share =
+            self.typ
+                .prove(encoded_measurement, &prove_rand, &joint_rand)?;
+
+        // Generate the proof shares.
+        for (x, y) in leader_proofs_share
+            .iter_mut()
+            .zip(self.derive_helper_proofs_share(&helper_seed))
+        {
+            *x -= y;
+        }
+
+        // Prep the output messages.
+        let leader_proof_share = SzkProofShare::Leader {
+            uncompressed_proof_share: leader_proofs_share,
+            blind_and_joint_rand_part: blind_and_joint_rand_part_for_leader,
+        };
+        let helper_proof_share = SzkProofShare::Helper {
+            proof_share_seed_and_blind: helper_seed.clone(),
+            joint_rand_part: helper_joint_rand_part,
+        };
+        Ok([leader_proof_share, helper_proof_share])
+    }
+
+    fn query(
+        &self,
+        input_share: &[T::Field],
+        proof: SzkProofShare<SEED_SIZE, T::Field>,
```

Comment:
I'd call this the `proof_share` and the proof share for the FLP below the "flp_proof_share".